modules/core/src/main/scala/iota/internal/toolbelts.scala

```diff
@@ -109,43 +111,101 @@ private[internal] sealed abstract class IotaCommonToolbelt {
-  final lazy val CopKTpe  = typeOf[iota.CopK[Nothing, Nothing]].etaExpand.resultType
+  private[this] def literalInt(tpe: Type): Either[Id[String], Int] =
+    tpe match {
+      case ConstantType(Constant(value: Int)) => value.asRight
+      case _ => s"Expected $tpe to be a literal integer".asLeft
+    }
+
+  class Parser private[IotaCommonToolbelt](
+    ConsSym   : Symbol,
+    NilSym    : Symbol,
+    ConcatSym : Symbol,
+    ReverseSym: Symbol,
+    TakeSym   : Symbol,
+    DropSym   : Symbol
+  ) {
+
+    final def apply(tpe: Type): Either[Id[String], Node] = tpe.dealias match {
+      case TypeRef(_, sym, args) =>
+        sym.asType.toType.dealias.typeSymbol match {
+          case ConsSym    => apply(args(1)).map(Cons(args(0), _))
+          case NilSym     => NNil.asRight
+          case ConcatSym  => apply(args(0)).map2(apply(args(1)))(Concat(_, _))
+          case ReverseSym => apply(args(0)).map(Reverse(_))
+          case TakeSym    => literalInt(args(0)).map2(apply(args(1)))(Take(_, _))
+          case DropSym    => literalInt(args(0)).map2(apply(args(1)))(Drop(_, _))
+          case sym        =>
+            s"Unexpected symbol $sym for type $tpe".asLeft
+        }
+      case ExistentialType(_, res) => apply(res)
+      case _ => s"Unable to parse type $tpe".asLeft
+    }
 
-  @tailrec
-  final def typeListFoldLeft[A](
-    nilTpe: Type, consTpe: Type
-  )(tpe: Type)(a0: A)(f: (A, Type) => A): Either[String, A] = tpe.dealias match {
-    case TypeRef(_, sym, Nil) if sym.asType.toType <:< nilTpe => Right(a0)
-    case TypeRef(_, sym, head :: tail :: Nil) if sym.asType.toType <:< consTpe =>
-      typeListFoldLeft(nilTpe, consTpe)(tail)(f(a0, head))(f)
-    case _ =>
-      Left(s"Unexpected type ${showRaw(tpe)} when inspecting type list")
   }
 
-  final def typeListTypes(
-    nilTpe: Type, consTpe: Type
-  )(tpe: Type): Either[Id[String], List[Type]] =
-    typeListFoldLeft(nilTpe, consTpe)(tpe)(List.empty[Type])((acc, t) => t :: acc).map(_.reverse)
+  private[this] def symbolOf[T](implicit evT: WeakTypeTag[T]): Symbol = evT.tpe.typeSymbol
+
+  final lazy val parseTList: Parser = new Parser(
+    NilSym     = symbolOf[iota.TNil],
+    ConsSym    = symbolOf[iota.TCons[Nothing, Nothing]],
+    ConcatSym  = symbolOf[iota.TList.Op.Concat[Nothing, Nothing]],
+    ReverseSym = symbolOf[iota.TList.Op.Reverse[Nothing]],
+    TakeSym    = symbolOf[iota.TList.Op.Take[Nothing, Nothing]],
+    DropSym    = symbolOf[iota.TList.Op.Drop[Nothing, Nothing]])
+
+  final lazy val parseKList: Parser = new Parser(
+    NilSym     = symbolOf[iota.KNil],
+    ConsSym    = symbolOf[iota.KCons[Nothing, Nothing]],
+    ConcatSym  = symbolOf[iota.KList.Op.Concat[Nothing, Nothing]],
+    ReverseSym = symbolOf[iota.KList.Op.Reverse[Nothing]],
+    TakeSym    = symbolOf[iota.KList.Op.Take[Nothing, Nothing]],
+    DropSym    = symbolOf[iota.KList.Op.Drop[Nothing, Nothing]])
+
+  final def evalTree(tree: Node): List[Type] = tree match {
```

Comment:
Same as above-- I would like to make this stack safe down the road, if possible.